src/dune_targets/dune_targets.ml

```diff
@@ -217,35 +217,54 @@ module Produced = struct
-    let rec collect (dir : Path.Build.t) (local : Path.Local.t)
-      : (unit Filename.Map.t Path.Local.Map.t, Error.t) result
-      =
-      match Path.readdir_unsorted_with_kinds (Path.build dir) with
-      | Error (Unix.ENOENT, _, _) -> Error (Missing_dir dir)
-      | Error e -> Error (Unreadable_dir (dir, e))
-      | Ok dir_contents ->
-        let open Result.O in
-        let+ filenames, dirs =
-          Result.List.fold_left
-            dir_contents
-            ~init:(Filename.Map.empty, Path.Local.Map.empty)
-            ~f:(fun (acc_filenames, acc_dirs) (filename, kind) ->
-              match (kind : File_kind.t) with
-              (* CR-someday rleshchinskiy: Make semantics of symlinks more consistent. *)
-              | S_LNK | S_REG ->
-                Ok (Filename.Map.add_exn acc_filenames filename (), acc_dirs)
-              | S_DIR ->
-                let+ dir =
-                  collect
-                    (Path.Build.relative dir filename)
-                    (Path.Local.relative local filename)
-                in
-                acc_filenames, Path.Local.Map.union_exn acc_dirs dir
-              | _ -> Error (Unsupported_file (Path.Build.relative dir filename, kind)))
-        in
-        if not (Filename.Map.is_empty filenames)
-        then Path.Local.Map.add_exn dirs local filenames
-        else dirs
+    let collect dir local =
+      let rec collect (dir : Path.Build.t) (local : Path.Local.t)
+        : (unit Filename.Map.t Path.Local.Map.t, Error.t) result
+        =
+        match Path.readdir_unsorted_with_kinds (Path.build dir) with
+        | Error (Unix.ENOENT, _, _) -> Error (Missing_dir dir)
+        | Error e -> Error (Unreadable_dir (dir, e))
+        | Ok dir_contents ->
+          let open Result.O in
+          let+ filenames, dirs =
+            Result.List.fold_left
+              dir_contents
+              ~init:(Filename.Map.empty, Path.Local.Map.empty)
+              ~f:(fun (acc_filenames, acc_dirs) (filename, kind) ->
+                match (kind : File_kind.t) with
+                (* CR-someday rleshchinskiy: Make semantics of symlinks more consistent. *)
+                | S_LNK | S_REG ->
+                  Ok (Filename.Map.add_exn acc_filenames filename (), acc_dirs)
+                | S_DIR ->
+                  let+ dir =
+                    collect
+                      (Path.Build.relative dir filename)
+                      (Path.Local.relative local filename)
+                  in
+                  acc_filenames, Path.Local.Map.union_exn acc_dirs dir
+                | _ -> Error (Unsupported_file (Path.Build.relative dir filename, kind)))
+          in
+          if not (Filename.Map.is_empty filenames)
+          then Path.Local.Map.add_exn dirs local filenames
+          else dirs
+      in
+      let rec parent_dirs path dirs =
+        match Path.Local.parent path with
+        | Some p when Path.Local.is_root p -> dirs
```

Comment:
Is it a problem that this doesn't stop at the root of the project/workspace?

Reply:
Or does it stop at the project root? I notice that the `Path.Local` module treats "." as the root so perhaps in practice this traversal never goes above the project root since the current directory will be somewhere in the project. If this is the case I think it deserves a comment explaining it, especially since my first instinct as a reader is to assume that `Path.Local.is_root` checks whether a path is "/" rather than ".".